Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,667 @@
+  `swift-service-lifecycle` takes care of listening for termination signals and terminating the services as well as the
+  `LambdaRuntime` in correct order.
+- `LambdaTerminator` can now be eliminated because its role is replaced with `swift-service-lifecycle`. The termination
+  logic of the Lambda function will be implemented in the conforming `run()` function of `LambdaRuntime`.
+
+With this, the earlier code snippet can be replaced with something much easier to read, maintain, and debug:
+
+```swift
+/// Instantiate services
+let postgresClient = PostgresClient()
+
+/// Instantiate LambdaRuntime with a closure handler implementing the business logic of the Lambda function
+let runtime = LambdaRuntime { (event: Input, context: LambdaContext) in
+    /// Use initialized service within the handler
+    try await postgresClient.query(...)
+}
+
+/// Use ServiceLifecycle to manage the initialization and termination
+/// of the services as well as the LambdaRuntime
+let serviceGroup = ServiceGroup(
+    services: [postgresClient, runtime],
+    configuration: .init(gracefulShutdownSignals: [.sigterm]),
+    logger: logger
+)
+try await serviceGroup.run()
+```
+
+### Simplifying Codable support
+
+A detailed explanation is provided in the **Codable Support** section. In short, much of the boilerplate code defined for
+each handler protocol in `Lambda+Codable` and `Lambda+String` will be replaced with a single `LambdaCodableAdapter`
+class.
+
+This adapter class is generic over (1) any handler conforming to `LambdaHandler`, (2) the user-specified input and
+output types, and (3) any decoder and encoder conforming to `LambdaEventDecoder` and `LambdaOutputDecoder`. The adapter
+will wrap the underlying handler with encoding/decoding logic.
+
+### Simplifying the handler protocols
+
+There are four different handler protocols in the current API (`SimpleLambdaHandler`, `LambdaHandler`,
+`EventLoopLambdaHandler`, `ByteBufferLambdaHandler`). As noted in the **Current Limitations** section, the ease-of-use
+varies for each handler protocol and users may not be able to easily determine which protocol best serves their use-case
+without spending time digging into the library. To reduce this problem and provide users with clear-cut options, we
+propose replacing all of the existing handler protocols with just two: `LambdaHandler` and `StreamingLambdaHandler`.
+Both will be explained in the **Detailed Solution** section.
+
+## Detailed Solution
+
+Below are explanations for all types that we want to use in AWS Lambda Runtime v2.
+
+### LambdaResponseWriter
+
+We will introduce a new `LambdaResponseWriter` protocol. It is used in the new `StreamingLambdaHandler` (defined below),
+which is the new base protocol for the `LambdaRuntime` (defined below as well).
+
+```swift
+/// A writer object to write the Lambda response stream into
+public protocol LambdaResponseWriter: ~Copyable {
```

Comment:
I think making this `~Copyable` will lead to a load of issues. Can we have a motivation on why we allow non-copyable types to implement this? If we allow non-copyable types here, then anything accepting `LambdaResponseWriter`s needs to support this which sounds like a tall ask.

Reply:
`LambdaResponseWriter` is not a protocol that users would need to implement. It is marked `NonCopyable` as we want to enforce through the compiler that users can `write()` after they have called `finish()`.

> If we allow non-copyable types here, then anything accepting `LambdaResponseWriter`s needs to support this which sounds like a tall ask.

Would you mind elaborating on this a bit?

Reply:
Sure, but don't you provide a `~Copyable` `LambdaResponseWriter` to the user?

Reply:
The semantics of the writer type we want is of mutable non-shared state. Isn't a non-copyable type the correct mechanism to express these semantics?